File: app/src/main/java/dev/eyosiyas/smsblocker/adapter/MessageAdapter.java
package dev.eyosiyas.smsblocker.adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.eyosiyas.smsblocker.R;
import dev.eyosiyas.smsblocker.model.Message;
import dev.eyosiyas.smsblocker.util.Core;
import dev.eyosiyas.smsblocker.view.DetailSmsActivity;

public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.ViewHolder> {
    private Context context;
    private List<Message> messages;

    public MessageAdapter(Context context, List<Message> messages) {
        this.context = context;
        this.messages = messages;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.sms_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final Message message = messages.get(position);
        if (!message.getSender().equals(message.getDisplayName()))
            holder.sender.setText(message.getDisplayName());
        else
            holder.sender.setText(message.getSender());
        holder.body.setText(message.getBody());
        holder.time.setText(Core.getReadableTime(message.getTimestamp()));
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                context.startActivity(new Intent(context, DetailSmsActivity.class).putExtra("KEY", message.getSender()));
            }
        });
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ImageView profile;
        TextView sender, body, time;

        public ViewHolder(@NonNull final View itemView) {
            super(itemView);
            profile = itemView.findViewById(R.id.UserProfile);
            sender = itemView.findViewById(R.id.txtSenderName);
            body = itemView.findViewById(R.id.txtMessageBody);
            time = itemView.findViewById(R.id.txtTime);
        }
    }
}

File: app/src/main/java/dev/eyosiyas/smsblocker/adapter/MessageAdapter.kt
package dev.eyosiyas.smsblocker.adapter

import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import com.amulyakhare.textdrawable.TextDrawable
import com.amulyakhare.textdrawable.util.ColorGenerator
import dev.eyosiyas.smsblocker.R
import dev.eyosiyas.smsblocker.databinding.SmsItemBinding
import dev.eyosiyas.smsblocker.event.MessageSelected
import dev.eyosiyas.smsblocker.model.Message
import dev.eyosiyas.smsblocker.util.Constant.DEFAULT_PROFILE
import dev.eyosiyas.smsblocker.util.Core


class MessageAdapter(private val selectedMessage: MessageSelected, private val messages: List<Message>) : RecyclerView.Adapter<MessageAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.sms_item, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setMessage(position)
    }

    override fun getItemCount(): Int {
        return messages.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val binder: SmsItemBinding = SmsItemBinding.bind(itemView)

        fun setMessage(position: Int) {
            val message = messages[position]
            val sender = if (message.sender != message.displayName) message.displayName else message.sender
            binder.txtSenderName.text = sender
            if (message.picture == DEFAULT_PROFILE) {
                val initial: Char = sender.toCharArray()[0]
                if (message.displayName != message.sender) {
                    if (initial.isLetter())
                        binder.UserProfile.setImageDrawable(TextDrawable.builder()
                                .buildRound(initial.toString(), ColorGenerator.MATERIAL.getColor(sender)))
                } else
                    binder.UserProfile.setImageDrawable(ResourcesCompat.getDrawable(itemView.context.resources, R.drawable.ic_user, null))
            } else
                binder.UserProfile.setImageURI(Uri.parse(message.picture))
            binder.txtMessageBody.text = message.body
            binder.txtTime.text = Core.interactiveTime(message.timestamp, itemView.context)
            itemView.setOnClickListener { selectedMessage.onMessageSelected(message, binder.txtSenderName.text.toString()) }
        }
    }
}